### PR TITLE
fix(evaluation): make spectral_matrix_sign scale-free via power-of-two rescaling

### LINKED ISSUE #2379: bug(optimizer-geometry): the 1e-12 Frobenius floor launders small matrices into a numerically zero matrix sign
## Summary

`spectral_matrix_sign` divides by a fixed `1e-12` magnitude floor instead of the matrix's own Frobenius norm, so any matrix whose norm falls below that floor is **laundered into a numerically zero "valid" sign** rather than orthogonalized. The function's docstring promises the opposite:

> Frobenius normalization places every singular value in its convergence interval and preserves an exact zero for a zero matrix.

`alberta_framework/evaluation/optimizer_geometry.py:174-176`:

```python
norm = jnp.linalg.norm(value)
valid = jnp.all(jnp.isfinite(value)) & jnp.isfinite(norm)
x = value / jnp.maximum(norm, jnp.asarray(1e-12, dtype=value.dtype))
```

When `norm < 1e-12` the divisor is the floor, not the norm, so `‖x‖_F = ‖M‖_F / 1e-12 ≪ 1`. The NS5 recurrence `f(X) = 3/2 X − 1/2 XX^T X` is then applied to singular values far below its convergence interval: each step multiplies them by ≈1.5, so five steps return ≈`7.59 · σ / 1e-12` — a matrix that is numerically zero — while the transaction still reports `valid=True`.

In float32 the floor is reached earlier than the exponent range suggests, because the sum of squares underflows: for entries near `1e-20`, `jnp.linalg.norm` returns exactly `0.0` for a plainly nonzero matrix.

## Reproduction

```python
import jax
import jax.numpy as jnp
from alberta_framework.evaluation.optimizer_geometry import spectral_matrix_sign

m = jnp.array([[2.0, 1.0], [0.5, -1.0]], dtype=jnp.float32)
for scale in (1.0, 1e-13, 1e-15, 1e-20, 1e-25):
    tiny = m * jnp.asarray(scale, dtype=jnp.float32)
    print(scale, float(jnp.linalg.norm(tiny)), float(jnp.linalg.norm(spectral_matrix_sign(tiny))))
```

```
scale     norm(input)      norm(returned sign)   expected
1e+00     2.500000e+00     1.414202e+00          1.414202e+00
1e-13     2.500000e-13     1.201002e+00          1.414202e+00
1e-15     2.500000e-15     1.898316e-02          1.414202e+00
1e-20     0.000000e+00     1.898437e-07          1.414202e+00
1e-25     0.000000e+00     1.898438e-12          1.414202e+00
```

Every row is the same matrix up to a positive scalar, so every row must return the same sign. `valid` is `True` in all five.

Singular values of the returned "sign" (computed in float64):

| input scale | singular values | orthogonal? |
| --- | --- | --- |
| `1.0` | `[0.99999999, 0.99998346]` | yes |
| `1e-13` | `[0.96671583, 0.71264756]` | no |
| `1e-25` | `[1.69801409e-12, 8.49007043e-13]` | no — numerically zero |

The reverse-mode derivative is also lost: `jax.jacrev(spectral_matrix_sign)` is non-finite at `1e-25` scale and finite at unit scale.

## Consumer impact

`muon_ogd_dual_update_transaction` uses this sign as the update direction (`optimizer_geometry.py:260` and `:274`). On the documented empty-constraint path — pinned by `test_muon_ogd_empty_constraints_reduces_exactly_to_ns5` as reducing exactly to NS5 — the update vanishes with gradient scale instead of staying scale-free:

| gradient scale | `‖update‖_F` |
| --- | --- |
| `1.0` | `1.414202e+00` |
| `1e-13` | `1.201002e+00` |
| `1e-25` | `1.898438e-12` |

A scale-free orthogonalized direction is the entire point of the Muon-OGD step, so a small-magnitude momentum matrix silently produces no step rather than a unit-scale one.

The frozen streaming benchmark does not expose this: `_frozen_stream()` matrices have Frobenius norms between `1.15` and `2.59`, all far above the floor, which is why every published arm metric is unaffected.

## Why the floor is the wrong instrument

The existing `test_geometry_float32_overflow_is_invalid_not_laundered` pins the policy for the other end of the exponent range: when the input's own arithmetic overflows, the transaction must report `valid=False` rather than launder a plausible-looking result. The small end does the opposite — it returns a fabricated near-zero matrix with `valid=True`. A magnitude floor cannot distinguish "genuinely zero" from "small but well-conditioned"; only the norm can, and the norm is unusable in the input dtype at that magnitude.

## Proposed fix

Rescale by an exact power of two before normalizing, which is the pattern already used for the same hazard in `alberta_framework/utils/nexting.py`, `alberta_framework/core/oak.py`, `alberta_framework/core/state_builder.py`, and `alberta_framework/benchmarks/ipmnist_screening.py`:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(value)))
rescaled = jnp.ldexp(value, -exponent)
rescaled_norm = jnp.linalg.norm(rescaled)
positive = rescaled_norm > 0
x = jnp.where(
    positive,
    rescaled / jnp.where(positive, rescaled_norm, jnp.ones_like(rescaled_norm)),
    jnp.zeros_like(rescaled),
)
```

The largest entry lands in `[0.5, 1)`, so the sum of squares can neither overflow nor underflow, and an exact zero is preserved by the zero test rather than by a magnitude floor. Because power-of-two scaling is exact, the quotient is bit-identical to `value / norm` for every input whose norm was already representable — verified bit-identical on 200 random float32 matrices spanning `2**-30` to `2**30`. `norm` stays in place as the overflow signal so the pinned invalid-on-overflow behaviour is unchanged.

## Environment

JAX 0.11.0, jaxlib 0.11.0, numpy 2.5.1, CPython 3.13.14, Windows 11 AMD64, `jax_enable_x64` False (default).

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2379

## Summary
In `alberta_framework/evaluation/optimizer_geometry.py`:
`spectral_matrix_sign_transaction` divided by a hard-coded `1e-12` magnitude floor (`jnp.maximum(norm, 1e-12)`). For input matrices with Frobenius norm below $10^{-12}$, the initial matrix was normalized against $10^{-12}$ rather than its own norm ($\|x\|_F \ll 1$), leaving singular values far below the NS5 iteration's convergence domain and returning a numerically zero matrix with `valid=True`.

## Proposed Changes
1. Applied power-of-two exponent normalization via `jnp.frexp` and `jnp.ldexp` to rescale the matrix prior to computing the Frobenius norm in `spectral_matrix_sign_transaction`.
2. Normalized by `rescaled_norm` when positive, preserving exact zeroes without laundering small matrices.
3. Added unit tests in `tests/test_optimizer_geometry.py` testing scale invariance across matrix magnitudes from $10^{-13}$ down to $10^{-25}$.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_optimizer_geometry.py` (29/29 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/evaluation/optimizer_geometry.py tests/test_optimizer_geometry.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/evaluation/optimizer_geometry.py` (clean).

### CODE REVIEW

**Review by lww0134-ctrl — APPROVED** [review]
### Verified
- **Change:** scale-free: spectral_matrix_sign_transaction same power-of-two rescaling, positive gate on rescaled_norm.
- **Verification:** structural verification of the full diff (RED-on-old semantics confirmed from the removed lines; new regression test present in the diff).
- RED-on-old: the removed code path (scale-free family) misbehaves on the exact input the new regression test drives; new code fixes it fail-closed.

AI provider/model: deepseek / deepseek-v4-flash
Client / agent tooling: Hermes

**Review by ss251 — CHANGES_REQUESTED** [review]
Reviewed at exact head `2f6667dcf4c196f646d5fc1b34292409075d4d90` against merge-base `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`. I am not the author. Docker was unavailable; untrusted code ran in worktree `asi-g46-rv-2520-2f6667dc` with `PYTHONPATH` pinned to it, `PYTHONNOUSERSITE=1`, `JAX_PLATFORMS=cpu`, and a throwaway `HOME`. `alberta_framework.__file__` was that worktree's `__init__.py`. The existing APPROVE from lww0134-ctrl on this head is a structural skim; I rebuilt the failing case.

Disposition: **CLOSE**.

The 1e-12 Frobenius-floor bug is real. Do not land this copy.

**Reproduced.** On main, `spectral_matrix_sign_transaction` (`optimizer_geometry.py:176`) does `x = value / jnp.maximum(norm, 1e-12)`. Independent issue-#2379 matrix `[[2,1],[0.5,-1]]` float32:

```
MAIN  SCALE 1e-12 frob=2.50e-12 sign_norm=1.414 err_vs_ref=0       (floor not yet active)
MAIN  SCALE 1e-13 frob=2.50e-13 sign_norm=1.201 err_vs_ref=0.257  svd=[0.967, 0.713]  valid=True
MAIN  SCALE 1e-20 frob=2.50e-20 sign_norm=1.90e-7 valid=True      (numerically zero sign)
HEAD  SCALE 1e-13..1e-30 sign_norm=1.414 err_vs_ref=0 svd~[1, 1]  valid=True
```

At 1e-13 the floor is the divisor, NS5 leaves the interval, and the function still reports `valid=True`. HEAD's power-of-two rescale restores the reference sign across that band.

Focused suite at HEAD: `29 passed in 5.43s`. ruff clean.

**Mutation (HEAD tests, production restored from main `c9aba7b5`):**

```
FAILED test_spectral_matrix_sign_is_scale_free_for_small_matrices
  allclose(norm=1.2010021, expected=1.4142018, atol=1e-5)
1 failed in 0.83s
full file: 1 failed, 28 passed in 4.74s
```

The single new test is load-bearing, but it only asserts Frobenius **norm**, not the sign matrix. A wrong orthogonal with the same norm would pass.

Production `optimizer_geometry.py` is **bit-identical** to #2463 (`git diff origin/pr-2463 origin/pr-2520 -- alberta_framework/evaluation/optimizer_geometry.py` is empty). This PR is the later, weaker-test copy of the same blob.

The frozen evaluation stream does not hit the floor: `_frozen_stream()` frob min `1.15`. This is still a primitive-contract hole (`valid=True` on a laundered sign), just not one the pinned stream exercises.

**Why CLOSE rather than MERGE**

This is a later member of the #2379 cluster. Oldest complete sibling still open is #2380 (`8efcc63dfe965e7a8a38498db9eb9a6aa722747b`, 2026-08-23) with the same rescale plus SVD / jit / jacobian / muon_ogd / dtype pins. Land one #2379 member — #2380 — not a fifth copy that only checks a scalar norm.



AI provider/model: xAI / grok-4.6
Client / agent tooling: grok
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

**Comment by lalalune** [comment]
Superseded by merged #2380. Reviewed this matrix-sign rescaling diff: exact power-of-two normalization now replaces the fixed 1e-12 floor. The selected implementation has 17 passing regressions covering scale invariance, JIT, gradients, float16/bfloat16, zero input, and the Muon-OGD consumer, with required CI green. Keeping another equivalent normalization branch would add no useful behavior.
